AP-3243

chiichang opened the pull request:
support CustomCalendar function for APMLog

chiichang committed:
AP-3243

support CustomCalendar function for APMLog

chiichang committed:
AP-3243-PD

added the missing code for supporting CustomCalendar at PD's stats UI

brucenguyen1 commented on Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/logobjects/AbstractLogImpl.java:
```diff
@@ -101,4 +104,8 @@ protected void setActivityInstances(List<ActivityInstance> activityInstances) {
         this.timeZone = timeZone;
     }
+
+    public void setCalendarModel(CalendarModel calendarModel) {
+        this.calendarModel = calendarModel;
+    }
```
It's better to add a Lombok's @NonNull constraint, e.g. setCalendarModel(@NonNull CalendarModel calendarModel). It will help the debugging of NullPointerException because it will be thrown right in this set method if it happens rather than waiting until some methods encounter null and throw it.

brucenguyen1 replied:
Another comment is it's safer to initialize CalendarModel in the APMLog constructor rather than via setCalendarModel method whiich is separated from the constructor and can be forgotten to call. Without a calendar model, all duration computation for the log won't work, that's the reason why it's safer to set calendar model in APMLog constructor  and check for non-null. With it initialized in the constructor, Once APMLog has been created, it is always safe to use its calendar model, no need to check for calendarModel != null condition in many places in the code.

raboczi replied:
If possible, marking `calendarModel` as `final` would allow the compiler to confirm that it's not changed anywhere other than in the constructor.

chiichang replied:
Adding CalendarModel to constructor or making it final are not applicable.

chiichang replied:
making calendarModel final is not applicable

brucenguyen1 replied:
That's ok. I remembered now APMLog is cached when it's not associated with any calendar.

brucenguyen1 commented on Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/filter/typefilters/PathFilter.java:
```diff
@@ -299,14 +300,18 @@ private static boolean haveSameAttributeValue(ActivityInstance act1, ActivityIns
             long st = act1.getStartTime();
             long et = act2.getEndTime();
-            duration = et > st ? et - st : 0;
+            duration = calendarModel != null ?
+                    Long.valueOf(calendarModel.getDuration(new long[]{st}, new long[]{et})[0]).doubleValue() :
```
Can we add a utility method, e.g. CalendarDuration.getDuration(CalendarModel calendarModel), add the above code to that method, and then use this method everywhere? It is to the above code duplicated in many places, and some places may forget to do it correctly leading to errors.

brucenguyen1 replied:
Can we use the simpler API provided in CalendarModel (below) rather than the above method with conversion to long array? 
- DurationModel getDuration(Long starDateTimeUnixTs, Long endDateTimeunixTs)
Can use DurationModel.getDuration.getNano() or DurationModel.getDuration.get(ChronoUnit.MILLIS).

chiichang replied:
DurationModel.getDuration.getNano() not applicable
DurationModel.getDuration.get(ChronoUnit.MILLIS) caused error

brucenguyen1 replied:
Pls use Duration.toMillis: https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html#toMillis--

chiichang committed:
apmlog-minor-improvement

- update according to comments

chiichang committed:
Merge branch 'development' into AP-3243

chiichang committed:
added missing import

chiichang committed:
added missing import

chiichang committed:
fixed error code

- reviewer suggested code caused error. changed the code back to original.
- added missing license header to a certain classes

brucenguyen1 committed:
Create a util class for calendar duration and use it in PathFilter.

chiichang committed:
Code_improvement_for_AP3243

chiichang committed:
avoid returning null for getCalendarModel

chiichang committed:
minor code improvement

chiichang committed:
removed code that caused UnitTest fail

chiichang committed:
Merge branch 'development' into AP-3243

brucenguyen1 committed:
Merge remote-tracking branch 'origin/development' into AP-3243

brucenguyen1 committed:
Use CalendarModel.ABSOLUTE_CALENDAR as a default one.
